openembedded-core-mbl/scripts/lib/mbl-signing/signing/pki/backends.py

```diff
@@ -0,0 +1,257 @@
+    """
+
+    @property
+    @abstractmethod
+    def server_url(self):
+        """URL of the signing backend server."""
+
+    @abstractmethod
+    def generate_root(self, *args, **kwargs):
+        """Generate a root CA."""
+
+    @abstractmethod
+    def generate_intermediate(self, *args, **kwargs):
+        """Generate an intermediate CA."""
+
+    @abstractmethod
+    def create_or_update_role(self, *args, **kwargs):
+        """
+        Create or update a role.
+
+        Roles are a set of parameters used to create the certificate.
+        These parameters include TTL, allowed common names, key usage etc.
+        """
+
+    @abstractmethod
+    def request_certificate(self, *args, **kwargs):
+        """Request an end entity certificate."""
+
+    @abstractmethod
+    def read_certificate(self, *args, **kwargs):
+        """Read a stored certificate from the backend."""
+
+    @abstractmethod
+    def list_certificates(self, *args, **kwargs):
+        """List certificates stored in the backend."""
+
+
+class VaultBackend(SigningBackend):
+    """
+    Implementation of the SigningBackend interface.
+
+    The backend used in this case is Hashicorp Vault.
+    """
+
+    def __init__(self, server_url=None):
+        """
+        Initialise the Vault client.
+
+        :params server_url str: Vault server url.
+        """
+        self._client = hvac.Client(url=server_url)
+        try:
+            self._client.sys.is_initialized()
+        except ConnectionError:
+            raise EnvironmentError("Vault sever is not running!") from None
+
+    @property
+    def mount_points(self):
+        """List of the mounted secret engines."""
+        return self._client.sys.list_mounted_secrets_engines()
+
+    @property
+    def server_url(self):
+        """Vault server url."""
+        return self._client.url
+
+    @server_url.setter
+    def server_url(self, new_url):
+        """Vault server url setter."""
+        self._client.url = new_url
+
+    def add_mount(self, new_mount, max_lease_ttl):
+        """
+        Add a new mounted secrets engine and tune it.
+
+        :params new_mount str: Name of the new mount.
+        :params max_lease_ttl str: Max TTL for certs issued by this mount.
+        """
+        self._client.sys.enable_secrets_engine(
+            backend_type="pki", path=new_mount
+        )
+        self._client.sys.tune_mount_configuration(
+            max_lease_ttl=max_lease_ttl, path=new_mount
+        )
+
+    def generate_root(self, common_name, **params):
+        """
+        Generate a root CA.
+
+        :params common_name str: Common name for the CA.
+        :params params dict: Optional config params to pass to Vault.
+        """
+        self._client.sys.enable_secrets_engine(
+            backend_type="pki", path=common_name
+        )
+        self._client.sys.tune_mount_configuration(
+            max_lease_ttl=params.get("ttl", "8750h"), path=common_name
```

Comment:
Does this mean the root CA will only last a year?  Does this need to be configurable?
Or is this just the mount to the secrets backend, if so why do we need a year?

Reply:
It's the max ttl the secrets backend is capable of. It is configurable, a year is the default value.

https://docs.python.org/3/library/stdtypes.html#dict.get